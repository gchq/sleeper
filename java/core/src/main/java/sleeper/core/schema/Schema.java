/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    private Schema(Builder builder) {
        rowKeyFields = validateRowKeys(builder.rowKeyFields);
        sortKeyFields = validateSortKeys(builder.sortKeyFields);
        valueFields = validateValueKeys(builder.valueFields);
        validateNoDuplicates(streamAllFields(rowKeyFields, sortKeyFields, valueFields));
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Creates an instance of this class using a schema file.
     *
     * @param  schemaPath  the path to the schema file
     * @return             an instance of this class
     * @throws IOException if an error occurs reading from the file
     */
    public static Schema load(Path schemaPath) throws IOException {
        return loadFromString(Files.readString(schemaPath));
    }

    /**
     * Creates an instance of this class using a JSON string.
     *
     * @param  schemaJson the JSON string representing a schema
     * @return            an instance of this class
     */
    public static Schema loadFromString(String schemaJson) {
        return new SchemaSerDe().fromJson(schemaJson);
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return getMappedFields(rowKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return getMappedFields(sortKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<String> getRowKeyFieldNames() {
        return getMappedFields(rowKeyFields, Field::getName);
    }

    public List<String> getSortKeyFieldNames() {
        return getMappedFields(sortKeyFields, Field::getName);
    }

    public List<String> getValueFieldNames() {
        return getMappedFields(valueFields, Field::getName);
    }

    public List<String> getAllFieldNames() {
        return getMappedFields(streamAllFields(), Field::getName);
    }

    private <T> List<T> getMappedFields(List<Field> fields, Function<Field, T> mapping) {
        return getMappedFields(fields.stream(), mapping);
    }

    private <T> List<T> getMappedFields(Stream<Field> fields, Function<Field, T> mapping) {
        return fields
                .map(mapping)
                .collect(Collectors.toUnmodifiableList());
    }

    public List<Field> getAllFields() {
        return streamAllFields().collect(Collectors.toUnmodifiableList());
    }

    /**
     * Streams all fields in the schema. This includes row key, sort key, and value fields.
     *
     * @return a stream of all fields
     */
    public Stream<Field> streamAllFields() {
        return streamAllFields(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Gets a field by name.
     *
     * @param  fieldName the name of the field
     * @return           a field, or an empty optional if the field does not exist
     */
    public Optional<Field> getField(String fieldName) {
        return streamAllFields()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    /**
     * Saves this schema to a file.
     *
     * @param  path        the path of the file
     * @throws IOException if an error occured when writing the file
     */
    public void save(Path path) throws IOException {
        Files.writeString(path, new SchemaSerDe().toJson(this));
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Builder to create a schema object.
     */
    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateValueKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        return makeImmutable(fields);
    }

    private static List<Field> makeImmutable(List<Field> fields) {
        return Collections.unmodifiableList(new ArrayList<>(fields));
    }

    private static Stream<Field> streamAllFields(
            List<Field> rowKeyFields, List<Field> sortKeyFields, List<Field> valueFields) {
        return Stream.of(rowKeyFields, sortKeyFields, valueFields)
                .flatMap(List::stream);
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
