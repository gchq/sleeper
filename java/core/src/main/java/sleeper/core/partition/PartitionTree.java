/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.TreeMap;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toUnmodifiableList;

/**
 * Represents a tree of partitions. It can be used to traverse or query the tree, e.g. to find all ancestors of a
 * partition, partitions that are either parents of the partition, or grandparents, or great-grandparents.
 */
public class PartitionTree {
    private final Map<String, Partition> idToPartition;
    private final Partition rootPartition;

    public PartitionTree(Collection<Partition> partitions) {
        this.idToPartition = new TreeMap<>();
        partitions.forEach(p -> this.idToPartition.put(p.getId(), p));
        List<Partition> rootPartitions = partitions.stream().filter(p -> null == p.getParentPartitionId()).collect(Collectors.toList());
        // There should be exactly one root partition.
        if (rootPartitions.size() != 1) {
            throw new IllegalArgumentException("There should be exactly one root partition, found " + rootPartitions.size());
        }
        this.rootPartition = rootPartitions.get(0);
    }

    /**
     * Retrieves IDs of partitions that were split from a given parent partition.
     *
     * @param  partitionId              the ID of the parent partition
     * @return                          the IDs of the child partitions
     * @throws IllegalArgumentException if the parent partition does not exist
     */
    public List<String> getChildIds(String partitionId) throws IllegalArgumentException {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getChildPartitionIds();
    }

    /**
     * Retrieves the partitions that were split from a given parent partition.
     *
     * @param  partitionId the ID of the parent partition
     * @return             the child partitions
     */
    public List<Partition> getChildPartitions(String partitionId) {
        List<Partition> outList = new ArrayList<>();
        getChildIds(partitionId).forEach(childId -> {
            outList.add(idToPartition.get(childId));
        });

        return outList;
    }

    /**
     * Retrieves IDs of all partitions that were split to result in the given partition. Starts with the most recent
     * parent, and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partitionId the ID of the partition to find ancestors of
     * @return             all IDs of the partition's ancestors
     */
    public List<String> getAllAncestorIds(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return ancestorsOf(idToPartition.get(partitionId)).map(Partition::getId).collect(toList());
    }

    /**
     * Retrieves all partitions that were split to result in the given partition. Starts with the most recent parent,
     * and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partitionId the ID of the partition to find ancestors of
     * @return             all of the partition's ancestors
     */
    public List<Partition> getAllAncestors(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return ancestorsOf(idToPartition.get(partitionId)).collect(toList());
    }

    /**
     * Streams through all partitions that were split to result in the given partition. Starts with the most recent
     * parent, and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partition the ID of the partition to find ancestors of
     * @return           all of the partition's ancestors
     */
    public Stream<Partition> ancestorsOf(Partition partition) {
        return Stream.iterate(getParent(partition),
                currentPartition -> currentPartition != null,
                currentPartition -> getParent(currentPartition));
    }

    /**
     * Streams through all partitions that are descendents of the given partition. Starts with the child partitions,
     * then finds their child partitions, and so on. Traverses depth first.
     *
     * @param  partition the partition
     * @return           the partition's descendents
     */
    public Stream<Partition> descendentsOf(Partition partition) {
        return partition.getChildPartitionIds()
                .stream().map(this::getPartition)
                .flatMap(child -> Stream.concat(Stream.of(child), descendentsOf(child)));
    }

    /**
     * Retrieves the parent of a partition.
     *
     * @param  partition the partition
     * @return           its parent, or null if it has no parent
     */
    public Partition getParent(Partition partition) {
        if (partition.getParentPartitionId() == null) {
            return null;
        }
        return idToPartition.get(partition.getParentPartitionId());
    }

    /**
     * Retrieves a partition by its unique ID.
     *
     * @param  partitionId the ID of the partition
     * @return             the partition
     */
    public Partition getPartition(String partitionId) {
        return idToPartition.get(partitionId);
    }

    public List<Partition> getAllPartitions() {
        return List.copyOf(idToPartition.values());
    }

    public List<Partition> getLeafPartitions() {
        return idToPartition.values().stream().filter(Partition::isLeafPartition).collect(toUnmodifiableList());
    }

    /**
     * Retrieves the leaf partition that includes the given key.
     *
     * @param  schema schema of the Sleeper table
     * @param  key    values for the key to find the leaf partition for
     * @return        the leaf partition
     */
    public Partition getLeafPartition(Schema schema, Key key) {
        // Sanity check key is of the correct length
        if (key.size() != schema.getRowKeyFields().size()) {
            throw new IllegalArgumentException("Key must match the row key fields from the schema (key was "
                    + key + ", schema has row key fields " + schema.getRowKeyFields() + ")");
        }
        // If root partition is a leaf partition then key must be in it
        if (rootPartition.isLeafPartition()) {
            return rootPartition;
        }

        return descend(schema, rootPartition, key);
    }

    private Partition descend(Schema schema, Partition currentNode, Key key) {
        // Get child partitions
        List<String> childPartitionIds = currentNode.getChildPartitionIds();
        List<Partition> childPartitions = new ArrayList<>();
        childPartitionIds.forEach(c -> childPartitions.add(idToPartition.get(c)));

        // Which child is the key in?
        Partition child = null;
        for (Partition partition : childPartitions) {
            if (partition.isRowKeyInPartition(schema, key)) {
                child = partition;
                break;
            }
        }
        if (null == child) {
            throw new IllegalArgumentException("Found key that was not in any of the child partitions: key " + key
                    + ", child partitions " + childPartitions);
        }
        if (child.isLeafPartition()) {
            return child;
        }
        return descend(schema, child, key);
    }

    public Partition getRootPartition() {
        return rootPartition;
    }

    /**
     * Retrieves the partition whose region includes both keys. Finds the nearest partition to each of the keys, with
     * the fewest steps up the partition tree.
     *
     * @param  schema schema of the Sleeper table
     * @param  a      values of the first key
     * @param  b      values of the second key
     * @return        the partition that includes both keys
     */
    public Partition getNearestCommonAncestor(Schema schema, Key a, Key b) {
        return getNearestCommonAncestor(getLeafPartition(schema, a), getLeafPartition(schema, b));
    }

    /**
     * Retrieves the partition whose region includes both descendent partitions. Finds the partition with the fewest
     * steps up the partition tree that is an ancestor for both.
     *
     * @param  a the first partition
     * @param  b the second partition
     * @return   the partition that is an ancestor of both
     */
    public Partition getNearestCommonAncestor(Partition a, Partition b) {
        if (a.getId().equals(b.getId())) {
            return a;
        }
        Set<String> ancestorsB = new HashSet<>(getAllAncestorIds(b.getId()));
        for (String ancestorA : getAllAncestorIds(a.getId())) {
            if (ancestorsB.contains(ancestorA)) {
                return getPartition(ancestorA);
            }
        }
        return getRootPartition();
    }

    /**
     * Traverses the partition tree visiting the leaves first. Proceeds in steps where you remove the current leaf
     * partitions and visit the new leaves.
     * <p>
     * The partitions are also ordered by the min and max of their ranges. Each time the tree is split, the partition
     * on the left/min side of the split will always be displayed first in this ordering.
     * <p>
     * This produces an ordering which is natural to read when you care the most about the leaf partitions, but you also
     * want to be able to read the rest of the tree in a predictable way.
     *
     * @return all partitions in the tree in leaves first order
     */
    public Stream<Partition> traverseLeavesFirst() {
        return traverseLeavesFirst(streamLeavesInTreeOrder().toList(), new HashSet<>(), Stream.empty());
    }

    /**
     * Traverses all leaf partitions by visiting every node in the partition tree. Starts at the root and goes left
     * first then right.
     *
     * @return all leaf partitions in order of their position in the tree
     */
    public Stream<Partition> streamLeavesInTreeOrder() {
        // Establish ordering by combining depth-first tree traversal with the ordering of child IDs on each partition.
        // This should ensure that partitions on the left/min side of a split will always come first in the order.
        return leavesInTreeOrderUnder(getRootPartition());
    }

    private Stream<Partition> leavesInTreeOrderUnder(Partition partition) {
        if (partition.isLeafPartition()) {
            return Stream.of(partition);
        }
        // Always follow left/min side first. Child partition IDs should be left/min then right/max.
        return partition.getChildPartitionIds().stream()
                .map(this::getPartition)
                .flatMap(this::leavesInTreeOrderUnder);
    }

    private Stream<Partition> traverseLeavesFirst(
            List<Partition> leaves, Set<String> prunedIds, Stream<Partition> partialTraversal) {

        // Prune the current leaves from the tree.
        // Tracking the pruned partitions creates a logical tree without needing to update the actual tree.
        leaves.stream().map(Partition::getId).forEach(prunedIds::add);

        // Find the partitions that are the new leaves of the tree after the previous ones were pruned.
        // Ensure the ordering is preserved, as the leaves were given in the correct order.
        List<Partition> nextLeaves = distinctParentsOf(leaves)
                .filter(parent -> prunedIds.containsAll(parent.getChildPartitionIds()))
                .collect(Collectors.toList());

        // Build traversal stream before recursive call, so it's tail-recursive
        Stream<Partition> traversal = Stream.concat(partialTraversal, leaves.stream());
        if (nextLeaves.isEmpty()) {
            return traversal;
        } else {
            return traverseLeavesFirst(nextLeaves, prunedIds, traversal);
        }
    }

    private Stream<Partition> distinctParentsOf(List<Partition> partitions) {
        return partitions.stream()
                .map(Partition::getParentPartitionId).filter(Objects::nonNull)
                .distinct().map(this::getPartition);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionTree that = (PartitionTree) o;
        return idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idToPartition);
    }

    @Override
    public String toString() {
        return "PartitionTree{" +
                "idToPartition=" + idToPartition +
                '}';
    }
}
