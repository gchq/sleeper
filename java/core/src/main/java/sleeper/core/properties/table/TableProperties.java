/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import sleeper.core.properties.PropertyGroup;
import sleeper.core.properties.SleeperProperties;
import sleeper.core.properties.SleeperPropertiesPrettyPrinter;
import sleeper.core.properties.SleeperPropertiesValidationReporter;
import sleeper.core.properties.SleeperPropertyIndex;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;
import java.util.stream.Collectors;

import static sleeper.core.properties.PropertiesUtils.loadProperties;
import static sleeper.core.properties.table.TableProperty.SCHEMA;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * Contains values of the properties to configure a Sleeper table.
 */
public class TableProperties extends SleeperProperties<TableProperty> {

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    /**
     * Creates a copy of the given table properties.
     *
     * @param  tableProperties the table properties
     * @return                 the copy
     */
    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    /**
     * Creates and validates an instance of this class with the given property values.
     *
     * @param  instanceProperties the instance properties
     * @param  properties         the property values
     * @return                    the table properties
     */
    public static TableProperties createAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    /**
     * Creates an instance with different property values than the given instance. Performs no validation, and uses the
     * same instance properties as the given instance.
     *
     * @param  tableProperties the original table properties
     * @param  newProperties   the property values
     * @return                 the new table properties
     */
    public static TableProperties recreateWithoutValidation(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    /**
     * Sets the schema of the Sleeper table.
     *
     * @param schema the schema
     */
    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return createPrettyPrinter(writer);
    }

    /**
     * Creates a printer to be used to display all table properties.
     *
     * @param  writer the writer to write to
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.builder()
                .properties(TableProperty.getAll(), TablePropertyGroup.getAll())
                .writer(writer).build();
    }

    /**
     * Creates a printer to be used to display table properties in a given group.
     *
     * @param  writer the writer to write to
     * @param  group  the group to display
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinterWithGroup(
            PrintWriter writer, PropertyGroup group) {
        return SleeperPropertiesPrettyPrinter.builder()
                .sortedProperties(TableProperty.getAll().stream()
                        .filter(property -> property.getPropertyGroup().equals(group))
                        .collect(Collectors.toList()))
                .writer(writer).hideUnknownProperties(true).build();
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
