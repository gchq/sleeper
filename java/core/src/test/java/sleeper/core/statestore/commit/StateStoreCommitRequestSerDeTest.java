/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.commit;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionTree;
import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.properties.testutils.FixedTablePropertiesProvider;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.LongType;
import sleeper.core.statestore.FileReferenceFactory;
import sleeper.core.statestore.transactionlog.log.TransactionBodyStore;
import sleeper.core.statestore.transactionlog.transaction.FileReferenceTransaction;
import sleeper.core.statestore.transactionlog.transaction.PartitionTransaction;
import sleeper.core.statestore.transactionlog.transaction.StateStoreTransaction;
import sleeper.core.statestore.transactionlog.transaction.TransactionType;
import sleeper.core.statestore.transactionlog.transaction.impl.AddFilesTransaction;
import sleeper.core.statestore.transactionlog.transaction.impl.ClearFilesTransaction;
import sleeper.core.statestore.transactionlog.transaction.impl.InitialisePartitionsTransaction;

import java.time.Instant;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstancePropertiesWithId;
import static sleeper.core.properties.testutils.TablePropertiesTestHelper.createTestTableProperties;
import static sleeper.core.schema.SchemaTestHelper.createSchemaWithKey;

public class StateStoreCommitRequestSerDeTest {

    Schema schema = createSchemaWithKey("key", new LongType());
    InstanceProperties instanceProperties = createTestInstancePropertiesWithId("instanceid");
    TableProperties tableProperties = createTestTableProperties(instanceProperties, schema);
    StateStoreCommitRequestSerDe serDe;

    @BeforeEach
    void setUp() {
        tableProperties.set(TABLE_ID, "933cd595");
        tableProperties.set(TABLE_NAME, "test.table");
        serDe = new StateStoreCommitRequestSerDe(new FixedTablePropertiesProvider(tableProperties));
    }

    @Test
    void shouldSerialiseCompactionCommitInS3() {
        // Given
        String key = createTransactionObjectKey(Instant.parse("2025-01-14T15:30:00Z"), "test-transaction");
        StateStoreCommitRequest commitRequest = createCommitRequestInS3(key, TransactionType.REPLACE_FILE_REFERENCES);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSerialiseInitialisePartitionsInS3() {
        // Given
        String key = createTransactionObjectKey(Instant.parse("2025-01-14T15:30:00Z"), "test-transaction");
        StateStoreCommitRequest commitRequest = createCommitRequestInS3(key, TransactionType.INITIALISE_PARTITIONS);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSerialiseInitialisePartitionsDirectly() {
        // Given
        PartitionTransaction transaction = new InitialisePartitionsTransaction(new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", 50L)
                .buildList());
        StateStoreCommitRequest commitRequest = createDirectCommitRequest(transaction);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSerialiseAddFilesDirectly() {
        // Given
        PartitionTree partitions = new PartitionsBuilder(schema).singlePartition("root").buildTree();
        FileReferenceTransaction transaction = AddFilesTransaction.fromReferences(List.of(
                fileFactory(partitions).rootFile("file1", 100),
                fileFactory(partitions).rootFile("file2", 200)));
        StateStoreCommitRequest commitRequest = createDirectCommitRequest(transaction);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSeraliseFileTransactionWithoutTableProperties() {
        // Given
        StateStoreCommitRequestSerDe serDe = StateStoreCommitRequestSerDe.forFileTransactions();
        StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create("test-table", new ClearFilesTransaction());

        // When
        String json = serDe.toJson(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
    }

    private String createTransactionObjectKey(Instant now, String uuid) {
        return TransactionBodyStore.createObjectKey(tableProperties.get(TABLE_ID), now, uuid);
    }

    private StateStoreCommitRequest createCommitRequestInS3(String objectKey, TransactionType transactionType) {
        return StateStoreCommitRequest.create(tableProperties.get(TABLE_ID), objectKey, transactionType);
    }

    private StateStoreCommitRequest createDirectCommitRequest(StateStoreTransaction<?> transaction) {
        return StateStoreCommitRequest.create(tableProperties.get(TABLE_ID), transaction);
    }

    private FileReferenceFactory fileFactory(PartitionTree partitions) {
        return FileReferenceFactory.from(instanceProperties, tableProperties, partitions);
    }

}
