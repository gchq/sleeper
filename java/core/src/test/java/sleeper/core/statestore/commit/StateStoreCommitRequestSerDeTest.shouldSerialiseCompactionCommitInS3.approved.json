{
  "tableId": "933cd595",
  "transactionType": "REPLACE_FILE_REFERENCES",
  "bodyKey": "933cd595/statestore/transactions/2025-01-14T15:30:00Z-test-transaction.json"
}
